src/helpers/coordinates.ts

```diff
@@ -43,6 +48,41 @@ function isCharADigit(char: string) {
   return char >= "0" && char <= "9";
 }
 
+// we limit the max column to 3 letters and max row to 7 digits for performance reasons
+export const MAX_COL = lettersToNumber("ZZZ");
+export const MAX_ROW = 9999998;
+
+export function consumeSpaces(chars: TokenizingChars) {
+  while (chars.current === " ") {
+    chars.advanceBy(1);
+  }
+}
+
+export function consumeLetters(chars: TokenizingChars) {
+  if (chars.current === "$") chars.advanceBy(1);
+  if (!chars.current || !isCharALetter(chars.current)) {
+    return -1;
+  }
+  let letterPart = 0;
+  while (chars.current && isCharALetter(chars.current)) {
+    const colIndex = charToNumber(chars.shift());
+    letterPart = letterPart * 26 + colIndex;
+  }
+  return letterPart;
+}
+
+export function consumeDigits(chars: TokenizingChars) {
+  if (chars.current === "$") chars.advanceBy(1);
+  if (!chars.current || !isCharADigit(chars.current)) {
+    return -1;
+  }
+  let num = 0;
+  while (chars.current && isCharADigit(chars.current)) {
```

Comment:
interestingly, this `chars.current` is really important for performance (even though it could be removed and the code would behave exactly the same way).

Without `chars.current`, the input of `isCharADigit` can be a char or `undefined`. With `chars.current` the inputs are always chars and it looks like the JS engine is able to inline or heavily optimize the function and it's ~3x faster